src/numeric_index.c

```diff
@@ -479,48 +483,76 @@ typedef struct {
   double value;
 } NumericRangeEntry;
 
-static int __cmd_docId(const void *p1, const void *p2) {
+static int cmpdocId(const void *p1, const void *p2) {
   NumericRangeEntry *e1 = (NumericRangeEntry *)p1;
   NumericRangeEntry *e2 = (NumericRangeEntry *)p2;
 
   return (int)e1->docId - (int)e2->docId;
 }
-void *NumericIndexType_RdbLoad(RedisModuleIO *rdb, int encver) {
-  if (encver != 0) {
+
+/** Version 0 stores the number of entries beforehand, and then loads them */
+static size_t loadV0(RedisModuleIO *rdb, NumericRangeEntry **entriespp) {
+  uint64_t num = RedisModule_LoadUnsigned(rdb);
+  if (!num) {
     return 0;
   }
 
-  NumericRangeTree *t = NewNumericRangeTree();
-  uint64_t num = RedisModule_LoadUnsigned(rdb);
+  *entriespp = array_newlen(NumericRangeEntry, num);
+  NumericRangeEntry *entries = *entriespp;
+  for (size_t ii = 0; ii < num; ++ii) {
+    entries[ii].docId = RedisModule_LoadUnsigned(rdb);
+    entries[ii].value = RedisModule_LoadDouble(rdb);
+  }
+  return num;
+}
+
+/** Version 0 stores (id,value) pairs, with a final 0 as a terminator */
+static size_t loadV1(RedisModuleIO *rdb, NumericRangeEntry **entriespp) {
+  NumericRangeEntry *entries = array_new(NumericRangeEntry, 1 << 16);
+  while (1) {
+    NumericRangeEntry cur;
+    cur.docId = RedisModule_LoadUnsigned(rdb);
+    if (!cur.docId) {
+      break;
+    }
+    cur.value = RedisModule_LoadDouble(rdb);
+    entries = array_append(entries, cur);
+  }
+  *entriespp = entries;
+  return array_len(entries);
+}
 
-  // we create an array of all the entries so that we can sort them by docId
-  NumericRangeEntry *entries = calloc(num, sizeof(NumericRangeEntry));
-  size_t n = 0;
-  for (size_t i = 0; i < num; i++) {
-    entries[n].docId = RedisModule_LoadUnsigned(rdb);
-    entries[n].value = RedisModule_LoadDouble(rdb);
-    n++;
+void *NumericIndexType_RdbLoad(RedisModuleIO *rdb, int encver) {
+  if (encver > NUMERIC_INDEX_ENCVER) {
+    return NULL;
+  }
+
+  NumericRangeTree *t = NewNumericRangeTree();
+  NumericRangeEntry *entries = NULL;
+  size_t numEntries = 0;
+  if (encver == 0) {
+    numEntries = loadV0(rdb, &entries);
+  } else {
+    numEntries = loadV1(rdb, &entries);
   }
```

Comment:
We should use an explicit if here and another else, then if we increase the encver in the future we will not load it in older versions